src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+    oscar_assure(F::ModuleGens)
+
+If fields of `F` from the OSCAR side are not defined, they
+are computed, given the Singular side.
+"""
+function oscar_assure(F::ModuleGens)
+  if !isdefined(F, :O)
+    F.O = [convert(F.F, singular_generators(F)[i]) for i=1:Singular.ngens(singular_generators(F))]
+  end
+end
+
+@doc Markdown.doc"""
+    singular_assure(F::ModuleGens)
+
+If fields of `F` from the Singular side are not defined, they
+are computed, given the OSCAR side.
+"""
+function singular_assure(F::ModuleGens)
+  if !isdefined(F, :S)
+    if length(F) == 0
+      singular_ring = base_ring(F.SF)
+      F.S = Singular.Module(singular_ring, Singular.vector(singular_ring, singular_ring(0)))
+      return 
+    end
+    F.S = Singular.Module(base_ring(F.SF), [convert(F.SF,x) for x = oscar_generators(F)]...)
+    return
+  end
+  #F[Val(:S), 1]
+end
+
+# i-th entry of module generating set (taken from Oscar side)
+getindex(F::ModuleGens, i::Int) = getindex(F, Val(:O), i)
+
+@doc Markdown.doc"""
+    singular_module(F::FreeMod)
+
+Create a Singular module from an OSCAR free module.
+"""
+function singular_module(F::FreeMod)
+  Sx = singular_ring(base_ring(F), keep_ordering=false)
+  return Singular.FreeModule(Sx, dim(F))
+end
+
+@doc Markdown.doc"""
+    convert(SF::Singular.FreeMod, m::FreeModElem)
+
+Convert an OSCAR free module element to the Singular side.
+"""
+function convert(SF::Singular.FreeMod, m::FreeModElem)
+  g = Singular.gens(SF)
+  e = SF()
+  Sx = base_ring(SF)
+  for (p,v) = m.coords
+    e += Sx(v)*g[p]
+  end
+  return e
+end
+
+@doc Markdown.doc"""
+    convert(F::FreeMod, s::Singular.svector)
+
+Convert a Singular vector to a free module element on the OSCAR side.
+"""
+function convert(F::FreeMod, s::Singular.svector)
+  pos = Int[]
+  values = []
+  Rx = base_ring(F)
+  R = base_ring(Rx)
+  for (i, e, c) = s
+    f = Base.findfirst(x->x==i, pos)
+    if f === nothing
+      push!(values, MPolyBuildCtx(base_ring(F)))
+      f = length(values)
+      push!(pos, i)
+    end
+    push_term!(values[f], R(c), e)
+  end
+  pv = Tuple{Int, elem_type(Rx)}[(pos[i], base_ring(F)(finish(values[i]))) for i=1:length(pos)]
+  return FreeModElem(sparse_row(base_ring(F), pv), F)
+end
+
+@doc Markdown.doc"""
+    FreeModuleHom{T1, T2} <: ModuleMap{T1, T2} 
+
+Data structure for morphisms where the domain is a free module (`FreeMod`).
+`T1` and `T2` are the types of domain and codomain respectively.
+`FreeModuleHom` is a subtype of `ModuleMap`.
+When computed, the corresponding matrix (via `matrix()`) and inverse isomorphism
+(in case there exists one) (via `inv()`) are cached.
+"""
+mutable struct FreeModuleHom{T1, T2} <: ModuleMap{T1, T2} 
+  matrix::MatElem
+  header::MapHeader
+  inverse_isomorphism::ModuleMap
+  Hecke.@declare_other
+
+  # generate homomorphism of free modules from F to G where the Array a contains the images of
+  # the generators of F
+  function FreeModuleHom{T,S}(F::FreeMod{T}, G::S, a::Vector{<:Any}) where {T, S}
```

Comment:
Writing `{<:Any}` is redundant (as always, please also get rid of it in other places)
```suggestion
  function FreeModuleHom{T,S}(F::FreeMod{T}, G::S, a::Vector) where {T, S}
```